IBX-4996: Storefront tutorial (#1910)
ezsystems/developer-documentation: docs/templating/layout/customize_storefront_layout.md

|`cart/component/add_to_cart/add_to_cart.html.twig`|"add to cart" element|
|`cart/component/summary/summary.html.twig`|cart summary|

#### Extend twig template

```html+twig
{% extends '@IbexaCart/themes/standard/cart/component/minicart/minicart.html.twig' %}

{% block content %}
    <img
        class="ibexa-store-minicart__image"
        src="{{ asset('bundles/ibexastorefront/img/icons/cart.svg') }}"
        alt="{{ 'minicart.icon.alt'|trans|desc('Cart') }}"
    />
    {{ parent() }}
{% endblock %}
```

> To avoid self-reference, ``@IbexaCart` is used instead of `@ibexadesign`.

Reviewer: ```suggestion
To avoid self-reference, `@IbexaCart` is used instead of `@ibexadesign`.
```

> Components out-of-the-box are not styled as you can freely customize them according to your needs.

Reviewer: ```suggestion
Built-in components are not styled, so you can freely customize them according to your needs.
```

> CSS classes can be easily added like every element from the base Twig which has its own attr object.

Reviewer: ```suggestion
You can add CSS classes from the base Twig which has its own `attr` object.
```

For example, if you want to add custom CSS classes to quantity input in Add to Cart component, use the following:

```html+twig
{% set quantity_input_attr = {
    class: 'ibexa-store-input ibexa-store-input--number ibexa-store-add-to-cart__quantity-input',
} %}
```

> Every element is also inside its own block so that it is easy to override it.

Reviewer: ```suggestion
Every element is also inside its own block so you can override the whole block.
```

#### Extending JavaScript

> In case of the JavaScript component, you should import original class and extend it:

Reviewer: ```suggestion
In case of the JavaScript component, you should import the original class and extend it:
```

```js
import Minicart from '@ibexa-cart/src/bundle/Resources/public/js/component/minicart';

export default class StorefrontMinicart extends Minicart {}
```

> The example below shows how to add a clear button support to the maincart:

Reviewer: ```suggestion
The example below shows how to add a "Clear" button support to the maincart:
```
Maincart upper- or lowercase?

```js
import Maincart from '@ibexa-cart/src/bundle/Resources/public/js/component/maincart';

export default class StorefrontMaincart extends Maincart {
    constructor(options) {
        super(options);

        this.clearCartBtn = this.container.querySelector('.ibexa-store-maincart__clear-cart-btn');

        this.onCartClear = this.onCartClear.bind(this);
    }

    attachStorefrontMaincartListeners() {
        this.clearCartBtn.addEventListener('click', this.onCartClear, false);
    }

    onCartClear() {
        this.cart.empty();
    }
}
```

### Main cart

> The base widget for the main cart view must be customized as out-of-the-box it consists only of the container with items. Each item consists of `<div>` wrappers with quantity input and remove item button. With customization you can add layout containers and items' data such as title, or price.

Reviewer: ```suggestion
You must customize the base widget for the main cart view, because out-of-the-box it consists only of the container with items.
Each item consists of `<div>` wrappers with quantity input and remove item button. With customization you can add layout containers and items' data such as title or price.
```

Available Twigs:

- `@IbexaCart/themes/standard/cart/component/maincart/maincart.html.twig`

with parameters:

>     - `attr`
>     - `item_template_attr`
>     - `items_container_attr`
>     - `item_template_params`
>     - `item_template_path`
>     - `net_price_template`

Reviewer: ```suggestion
- `attr`
- `item_template_attr`
- `items_container_attr`
- `item_template_params`
- `item_template_path`
- `net_price_template`
```
Can you add descriptions to the params?

- `@IbexaCart/themes/standard/cart/component/maincart/maincart_item.html.twig`

with parameters:

>     - `cart_entry_quantity`
>     - `item_attr`
>     - `quantity_input_attr`
>     - `remove_item_btn_attr`

Reviewer: ```suggestion
- `cart_entry_quantity`
- `item_attr`
- `quantity_input_attr`
- `remove_item_btn_attr`
```

and the same in lists below

JavaScript class:

- `@ibexa-cart/src/bundle/Resources/public/js/component/maincart`

### Add to Cart

Available Twig:

- `@IbexaCart/themes/standard/cart/component/add_to_cart/add_to_cart.html.twig`

with parameters:

    - `is_disabled`
    - `attr`
    - `product_code`
    - `quantity_input_attr`
    - `add_to_cart_btn_attr`

JavaScript class:

- `@ibexa-cart/src/bundle/Resources/public/js/component/summary`

> ### Mini cart

Reviewer: One word or two words?

Author: One word, my mistake.

|`cart/component/add_to_cart/add_to_cart.html.twig`|"add to cart" element|
|`cart/component/summary/summary.html.twig`|cart summary|

#### Extend Twig template

```html+twig
{% extends '@IbexaCart/themes/standard/cart/component/minicart/minicart.html.twig' %}

{% block content %}
    <img
        class="ibexa-store-minicart__image"
        src="{{ asset('bundles/ibexastorefront/img/icons/cart.svg') }}"
        alt="{{ 'minicart.icon.alt'|trans|desc('Cart') }}"
    />
    {{ parent() }}
{% endblock %}
```

To avoid self-reference, `@IbexaCart` is used instead of `@ibexadesign`.

Built-in components are not styled, so you can freely customize them according to your needs.

> You can add CSS classes from the base Twig which has its own `attr` object.

Reviewer 2: ```suggestion
You can add CSS classes to the base Twig using attributes' objects.
```

For example, if you want to add custom CSS classes to quantity input in Add to Cart component, use the following:

```html+twig
{% set quantity_input_attr = {
    class: 'ibexa-store-input ibexa-store-input--number ibexa-store-add-to-cart__quantity-input',
} %}
```

Every element is also inside its own block so you can override the whole block.

#### Extending JavaScript

In case of the JavaScript component, you should import the original class and extend it:

```js
import Minicart from '@ibexa-cart/src/bundle/Resources/public/js/component/minicart';

export default class StorefrontMinicart extends Minicart {}
```

> The example below shows how to add a "Clear" button support to the maincart:

Reviewer 2: We should probably mention, that this example does not cover twig part, which is also required.
